src/utils/mtev_lfu.c

```diff
@@ -234,21 +231,27 @@ touch_lfu_cache_no_lock(mtev_lfu_t *c, struct lfu_entry *e)
 {
   if (c->max_entries > 0) {
     struct lfu_cache_entry *bucket = e->frequency_list_head;
-    struct lfu_cache_entry *next_bucket = STAILQ_NEXT(bucket, list_entry);
+    struct lfu_cache_entry *next_bucket = TAILQ_NEXT(bucket, list_entry);
+    struct lfu_cache_entry *prev_bucket = TAILQ_PREV(bucket, frequency_list, list_entry);
     size_t bucket_freq = bucket->frequency;
     mtev_boolean need_new_bucket = mtev_false;
 
     if (next_bucket == NULL ||
-        next_bucket == STAILQ_FIRST(&c->lfu_frequency_list) ||
+        next_bucket == TAILQ_FIRST(&c->lfu_frequency_list) ||
         next_bucket->frequency != bucket_freq + 1) {
       need_new_bucket = mtev_true;
     }
     struct lfu_cache_entry *empty = remove_from_frequency_list_no_lock(c, e);
     if (need_new_bucket) {
       if (empty != NULL) {
+        /* reuse this bucket instead of allocating */
+        empty->frequency = bucket_freq + 1;
```

Comment:
There may be gaps in the bucket list.

`1 -> 3 -> 7`

If we are in `3` and `3` becomes empty.. when it becomes 4 it needs to be inserted *after* `1`, not at the HEAD which would create the wrong order.

Now we check next and prev buckets.. if we have advanced past `next_bucket` then we treat it as the prior, if not, we treat `prev_bucket` as prior.  `prev_bucket` will always be lower in frequency by definition.